imap_processing/mag/l1a/mag_l1a_data.py

```diff
@@ -429,6 +479,478 @@ def to_signed16(n: int) -> int:
+        the number of vectors in vectors.
+
+        Parameters
+        ----------
+        range_data : numpy.ndarray
+            Array of range values, where each value is one bit.
+        vectors : numpy.ndarray
+            Array of vectors, where each vector is a tuple of (x, y, z, range).
+            The range value will be overwritten by range_data, and x, y, z will remain
+            the same.
+
+        Returns
+        -------
+        numpy.ndarray
+            Updated array of vectors, identical to vectors with the range values
+            updated from range_data.
+        """
+        if len(range_data) != (len(vectors) - 1) * 2:
+            raise ValueError(
+                "Incorrect length for range_data, there should be two bits per vector, "
+                "excluding the first."
+            )
+
+        updated_vectors: np.ndarray = np.copy(vectors)
+        range_str = "".join([str(i) for i in range_data])
+        for i in range(len(vectors) - 1):
+            range_int = int(range_str[i * 2 : i * 2 + 2], 2)
+            updated_vectors[i + 1][3] = range_int
+        return updated_vectors
+
+    @staticmethod
+    def accumulate_vectors(
+        first_vector: np.ndarray,
+        vector_differences: list[int],
+        vector_count: int,
+    ) -> np.ndarray:
+        """
+        Given a list of differences and the first vector, return calculated vectors.
+
+        This is calculated as follows:
+            vector[i][0] = vector[i-1][0] + vector_differences[i][0]
+            vector[i][1] = vector[i-1][1] + vector_differences[i][1]
+            vector[i][2] = vector[i-1][2] + vector_differences[i][2]
+            vector[i][3] = first_vector[3]
+
+        The third element of the array is the range value, which we assume is the same
+        as the first vector.
+
+        Parameters
+        ----------
+        first_vector : numpy.ndarray
+            A numpy array of 3 signed integers and a range value, representing the
+            start vector.
+        vector_differences : numpy.ndarray
+            A numpy array of shape (expected_vector_count, 4) of signed integers,
+            representing the differences between vectors.
+        vector_count : int
+            The expected number of vectors in the output.
+
+        Returns
+        -------
+        numpy.ndarray
+            A numpy array of shape (expected_vector_count, 4) of signed integers,
+            representing the calculated vectors.
+        """
+        vectors: np.ndarray = np.empty((vector_count, 4), dtype=np.int32)
+        vectors[0] = first_vector
+        if len(vector_differences) % 3 != 0:
+            logger.error(f"Error! Computed compressed vector differences are not "
+                         f"divisible by 3 - meaning some data is missing. "
+                         "Expected length: %s, actual length: "
+                         "%s", vector_count * 3, len(vector_differences))
```

Comment:
If this happened we would probably say abandon the packet as something very bad has happened

Reply:
I emailed you about this - this was originally an error, but I changed it to a log because the test packets you sent me had several packets where this error cropped up. 

In the future, would you want this to generate a CDF file that is empty, or skip generating the CDF file altogether? Skip that packet only, but include the other packets if they're ok? Pass in the data we have but mark a quality flag? Something else....?

Thanks for your review!